Add transaction protos

shannynalayna opened the pull request:
Creates the bond transaction protos based on the current state objects.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

shannynalayna commented on protos/bond_payload.proto:
```diff
@@ -0,0 +1,47 @@
+	Action action = 1;
+
+	// Payload data
+	bytes content = 2;
```
Based this format on the Sawtooth NEXT protos payload file

shannynalayna commented on protos/holding_transaction.proto:
```diff
@@ -0,0 +1,44 @@
+	string owner_id = 1;
+
+	// Whether the asset is currency or a bond
+	Organization.Holding.AssetType asset_type = 2;
```
Slightly unsure of how to correctly access submessages, however, when building, the Organization prefix was necessary. Same issue for receipt_transaction.

peterschwarz replied:
I think this is correct.

shannynalayna commented on protos/receipt_transaction.proto:
```diff
@@ -0,0 +1,36 @@
+message CreateReceipt {
+
+	// Whether it was a coupon or redemption payment
+	Organization.Receipt.PaymentType payment_type = 1;
```
Slightly unsure of how to correctly access submessages, however, when building, the Organization prefix was necessary. Same issue for holding_transaction.

eloaverona commented on protos/holding_transaction.proto:
```diff
@@ -0,0 +1,44 @@
+message CreateHoldingAction {
+
+	// Unique identifier of the organization which owns this holding
+	string owner_id = 1;
```
I wonder if we need this field? Because this should be the id for the organization of the participant who submitted the transaction. I don't know if it makes a huge different either way.

shannynalayna replied:
True, the information of the organization which has the asset would be accessible through the participant that signed the transaction. Seems a bit redundant

eloaverona commented on protos/holding_transaction.proto:
```diff
@@ -0,0 +1,44 @@
+message DeleteHoldingAction {
+
+	// Unique identifier of the organization which owns this holding
+	string owner_id = 1;
```
Same as above, not sure if we need this field.

eloaverona commented on protos/order_transaction.proto:
```diff
@@ -0,0 +1,52 @@
+	Order.OrderType order_type = 2;
+
+	// Unique name of the organization the order is behalf on
+	string ordering_organization_id = 3;
```
Again, I wonder if we need this field? As this should be the id for the organization of the participant who submitted the transaction.

eloaverona commented on protos/participant_transaction.proto:
```diff
@@ -0,0 +1,42 @@
+    
+	// The public key associated with the participant used to sign
+	// transactions
+	string public_key = 1;
```
Would  participants be able to update information on other participants?  If not, I don't think we need the public_key here, as a participant would only be able to update their own info.

eloaverona commented on protos/quote_transaction.proto:
```diff
@@ -0,0 +1,37 @@
+message CreateQuoteAction {
+
+    // Unique identifier of the organization the quote is on behalf of
+    string organization_id = 1;
```
Redundant as well

eloaverona commented:
The organization transaction protos don't seem to be included.

eloaverona commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,96 @@
+	// The industry of the organization
+	string industry = 3;
+
+	// A code representing the identity of the organization
```
Is this always the ticker symbol? If yes, the comment could reflect that, for clarity.

eloaverona commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,96 @@
+	string industry = 3;
+
+	// A code representing the identity of the organization
+	string organization_id = 4;
```
I imagine the ticker symbol of a organization does not change? If it does not, then we don't need this field, as we can get the organization id from the participant who submitted the transaction.

eloaverona commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	// The name of the organization
+	string name = 1;
```
Is this field necessary?

eloaverona commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,96 @@
+message DeleteOrganizationAction {
+
+	// The name of the organization
+	string name = 1;
```
Is this field necessary? If a participant submits this transaction, we can just assume they want to delete their own org?

shannynalayna commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,88 @@
+message UpdateOrganizationAction {
+
+	// The name of the organization
+	string name = 1;
```
Similarly to the other org_ids that were unnecessary, presumably this is also redundant as I'm unsure if an org's name would change

eloaverona replied:
Yeah.. I don't know if org's name change either. But maybe?  I feel it's fine to leave name here for now.

eloaverona commented on protos/bond_transaction.proto:
```diff
@@ -0,0 +1,61 @@
+	string bond_id = 1;
+
+	// Ticker symbol of the organization which issued the bond
+	string issuing_organization_id = 2;
```
Would this also be redundant? It could be based on the organization id of the participant who created the transaction.

shannynalayna replied:
Yeah, we just have to make this a required field for participants, then. In the original implementation, that field was "optional" within the participant object

eloaverona replied:
Yeah, it cannot be required because a participant is created before a organization, right? But, the field should be filled  if participants want to submit certain transactions.

eloaverona commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,88 @@
+	string pricing_source = 4;
+
+	// List of the assets that the organization owns
+	repeated Organization.Holding holdings = 5;
```
I wonder, if maybe this field should not be included in the `UpdateOrganizationAction`? Because the transactions updates in relation to holdings are `CreateHoldingAction` and `DeleteHoldingAction`, which you already implemented elsewhere.

shannynalayna replied:
Wait, I misunderstood your comment. Yeah, that makes sense. And, it seems like the other transactions are carrying the bond/currency amounts to update the specific org's holdings.

dplumb94 commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,85 @@
+
+import "organization.proto";
+
+message CreateOrganizationAction {
```
It looks like some messages have the `Action` prefix and some do not. I think you can just take it off of all of them.

dplumb94 commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,85 @@
+}
+
+message DeleteOrganizationAction {
+
```
Blank message

dplumb94 commented on protos/holding_transaction.proto:
```diff
@@ -0,0 +1,38 @@
+syntax = "proto3";
+
+import "organization.proto";
+
```
I think we need an organization ID for these transactions, since organizations may have holdings of the same bond or USD

shannynalayna replied:
Based on the github message, we can skip adding the org_id field to these transactions?

dplumb94 commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,81 @@
+    repeated Organization.Authorization authorizations = 6;
+
+    // List of the assets that the organization owns
+    repeated Organization.Holding holdings = 7;
```
Holdings are created in their own transaction, so we wont need a holdings field in the CreateOrganization txn

dplumb94 commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,83 @@
+
+message CreateOrganization {
+	
+    // The name of the organizaiton
```
typo

eloaverona commented on protos/organization_transaction.proto:
```diff
@@ -0,0 +1,83 @@
+    string organization_id = 2;
+
+    // Type of the organization
+    Organization.OrganizationType organization_type = 3;
```
If we are using the ticker or pricing source as id, depending on the organization type. We should not allow organization type to be updated, should we?

shannynalayna replied:
I was thinking the same thing when I went through and updated this portion... Maybe we should discuss implementing the update organization transaction seeing as how a lot of the information to how we identify the org is alterable based on this transaction.

eloaverona commented on protos/bond_payload.proto:
```diff
@@ -0,0 +1,46 @@
+    enum Action {
+        CREATE_ORGANIZATION = 0;
+        UPDATE_ORGANIZATION = 1;
+        UPDATE_ORGANIZATION_AUTHORIZATION = 2;
```
It seems the DeleteOrganization action is missing?

shannynalayna replied:
After taking out the organization_id in the DeleteOrganization message, I also took out that empty message and the associated action. Darian and I discussed the ability to have an action in the payload file without an actual message (which, if we did implement the delete org transaction we would probably need the org_id field in that message) so we thought it wasn't really necessary to implement that transaction.

eloaverona replied:
For me it makes sense to have the action without the message (if it is possible). I thought the reason we deleted the org_id in the DeleteOrganization message, was because we could determine the organization being deleted through the participant who submitted the transaction.

dplumb94 replied:
In 0.7 Bond, an organization can only be deleted by the organization creator. The creator is not tracked in this new version, so it is not clear what the logic should be for deleting an org. We could either not allow orgs to be deleted, or allow anyone to delete them. Not sure if it makes sense to tie this to a participant's currently affiliated organization.

Commit:
Fix quote's attribute description

Fixes state object quotes's attribute description for bid-price and
bid-qty to correctly reflect the description in the docs.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

eloaverona commented:
Some changes requested on the commit ["Add transaction protos"](https://github.com/bitwiseio/sawtooth-bond/pull/4/commits/78ee0a46e9f0c7e23e9edcb8d5a712bfd82595cd)  were fixed only on the following commit  ["Eliminate pricing source attribute in organization"](https://github.com/bitwiseio/sawtooth-bond/pull/4/commits/d699665914e19d4796e9bcf6f8bd27f1d59b4bd7)
The changes should be made in the commit that introduced the error.

Commit:
Add transaction protos

Creates the bond transaction protos based on state objects.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

Commit:
Eliminate pricing source attribute in organization

Eliminates the pricing source attribute from the organization object.
Consolidates this field into the organization_id attribute.

Signed-off-by: Shannyn Telander <telander@bitwise.io>

eloaverona commented on protos/quote_transaction.proto:
```diff
@@ -0,0 +1,36 @@
+    // The maximum quantity the firm is willing to sell.
+    uint64 ask_qty = 3;
+
+    // The maximum price for which the firm is willing to sell the bond.
```
In the comment, it should be buy, not sell.

eloaverona commented on protos/quote_transaction.proto:
```diff
@@ -0,0 +1,36 @@
+    // Format: Millionths of dollars
+    uint64 bid_price = 4;
+
+    // The minimum quantity the firm is willing to sell.
```
In the comment, it should be buy, not sell.

shannynalayna commented:
@eloaverona updated

peterschwarz reviewed (changes requested):
It all looks good, from a content perspective, but I think the files should be renamed.  If we think about them as the resulting packages or namespaces, we should be sure that they are organized in a predictable way. 

I would suggest renaming to match the payload term

- `bond_transaction.proto` -> `payload_bond.proto`
- `order_transaction.proto` -> `payload_order.proto`
-  etc

Thus the package names would be:

`payload`
`payload_bond`
`payload_order`
etc

shannynalayna commented:
@peterschwarz file names are updated
